src/yunohost/backup.py

```diff
@@ -49,111 +51,199 @@
+            self._hook_files(method, output_directory)
+            logger.info(m18n.n('backup_method_custom_finished', method=method))
+
+    def clean(self, retcode=-1):
+        """ Call post_backup_create hooks and delete collect_dir """
+        ret = hook_callback('post_backup_create', args=[self.collect_dir,
+                                                        retcode])
         if not ret['failed']:
-            filesystem.rm(tmp_dir, True, True)
+            if self._is_temp_collect_dir:
+                filesystem.rm(self.collect_dir, True, True)
             return True
         else:
             logger.warning(m18n.n('backup_cleaning_failed'))
             return False
 
-    # Create temporary directory
-    if not tmp_dir:
-        tmp_dir = "%s/tmp/%s" % (BACKUP_PATH, name)
-        if os.path.isdir(tmp_dir):
+    @property
+    def _is_temp_collect_dir(self):
+        return self.collect_dir == os.path.join(BACKUP_PATH, self.name)
+
+    def _define_backup_name(self):
+        """ Define backup name """
+        # FIXME: case where this name already exist
+        return time.strftime('%Y%m%d-%H%M%S')
+
+    def _init_collect_dir(self):
+        """ Initialize preparation directory """
+
+        if not self.collect_dir:
+            self.collect_dir = "%s/tmp/%s" % (BACKUP_PATH, name)
+
+        if not os.path.isdir(self.collect_dir):
+            filesystem.mkdir(self.collect_dir, 0750, parents=True, uid='admin')
+        elif self.bindable:
             logger.debug("temporary directory for backup '%s' already exists",
-                tmp_dir)
-            if not _clean_tmp_dir():
+                         self.collect_dir)
+            if not self.clean():
                 raise MoulinetteError(
                     errno.EIO, m18n.n('backup_output_directory_not_empty'))
-        filesystem.mkdir(tmp_dir, 0750, parents=True, uid='admin')
 
-    # Initialize backup info
-    info = {
-        'description': description or '',
-        'created_at': timestamp,
-        'apps': {},
-        'hooks': {},
-    }
+    def _open_csv(self):
+        """ Initialize backup list """
+        self.csv_path = os.path.join(self.collect_dir, 'backup.csv')
+        try:
+            self.csv_file = open(self.csv_path, 'a')
+            self.fieldnames = ['source', 'dest']
+            self.csv = csv.DictWriter(self.csv_file, fieldnames=self.fieldnames,
+                                      quoting=csv.QUOTE_ALL)
+        except (IOError, OSError, csv.Error):
+            logger.error(m18n.n('backup_csv_creation_failed'))
+
+    def _get_env_var(self):
+        """ Define environment variable for hooks call """
+        env_var = {}
+
+        _, tmp_csv = tempfile.mkstemp(prefix='backupcsv_')
+        env_var['YNH_BACKUP_DIR'] = self.collect_dir
+        env_var['YNH_BACKUP_CSV'] = tmp_csv
+        return env_var
+
+    def _commit_path_in_csv(self, tmp_csv):
```

Comment:
I have implemented this after some fail tests.
The goal of this function is to recopy the csv produce by a hook or a backup script of an app into the true CSV. Like that if the script fail, we don't backup the listed files.

But may be it's not the good behaviour and we should register path to backup even if the script fail...